internal/graph/singleflight_resolver_test.go

```diff
@@ -0,0 +1,371 @@
+
+	singleflightCheckResolver.SetDelegate(localCheckResolver)
+	localCheckResolver.SetDelegate(singleflightCheckResolver)
+
+	t.Cleanup(func() {
+		goleak.VerifyNone(t)
+		singleflightCheckResolver.Close()
+		localCheckResolver.Close()
+	})
+
+	for i := 0; i < 1000; i++ {
+		resp, err := singleflightCheckResolver.ResolveCheck(ctx, &ResolveCheckRequest{
+			StoreID:              storeID,
+			AuthorizationModelID: model.GetId(),
+			TupleKey:             tuple.NewTupleKey("document:1", "viewer", "user:anne"),
+			ResolutionMetadata: &ResolutionMetadata{
+				Depth: 25,
+			},
+		})
+		require.NoError(t, err)
+		require.True(t, resp.GetAllowed())
+	}
+}
+
+func TestSingleflightCheckResolver_ThreeProngLoop(t *testing.T) {
+	storeID := ulid.Make().String()
+
+	ds := memory.New()
+	err := ds.Write(context.Background(), storeID, nil, []*openfgav1.TupleKey{
+		tuple.NewTupleKey("module:a", "owner", "user:anne"),
+		tuple.NewTupleKey("folder:a", "parent", "module:a"),
+		tuple.NewTupleKey("document:a", "parent", "folder:a"),
+		tuple.NewTupleKey("module:b", "parent", "document:a"),
+		tuple.NewTupleKey("folder:b", "parent", "module:b"),
+		tuple.NewTupleKey("document:b", "parent", "folder:b"),
+		tuple.NewTupleKey("module:a", "parent", "document:b"),
+	})
+	require.NoError(t, err)
+
+	model := testutils.MustTransformDSLToProtoWithID(`model
+  schema 1.1
+
+type user
+type module
+relations
+	define owner: [user] or owner from parent
+	define parent: [document, module]
+	define viewer: [user] or owner or viewer from parent
+type folder
+relations
+	define owner: [user] or owner from parent
+	define parent: [module, folder]
+	define viewer: [user] or owner or viewer from parent
+type document
+relations
+	define owner: [user] or owner from parent
+	define parent: [folder, document]
+	define viewer: [user] or owner or viewer from parent`)
+
+	err = ds.WriteAuthorizationModel(context.Background(), storeID, model)
+	require.NoError(t, err)
+
+	ctx := typesystem.ContextWithTypesystem(
+		context.Background(),
+		typesystem.New(model),
+	)
+	ctx = storage.ContextWithRelationshipTupleReader(ctx, ds)
+
+	singleflightCheckResolver := NewSingleflightCheckResolver()
+	localCheckResolver := NewLocalChecker()
+
+	singleflightCheckResolver.SetDelegate(localCheckResolver)
+	localCheckResolver.SetDelegate(singleflightCheckResolver)
+
+	t.Cleanup(func() {
+		goleak.VerifyNone(t)
+		singleflightCheckResolver.Close()
+		localCheckResolver.Close()
+	})
+
+	tupleKeys := []*openfgav1.TupleKey{
+		tuple.NewTupleKey("module:a", "viewer", "user:anne"),
+		tuple.NewTupleKey("module:b", "viewer", "user:anne"),
+		tuple.NewTupleKey("folder:a", "viewer", "user:anne"),
+		tuple.NewTupleKey("folder:b", "viewer", "user:anne"),
+		tuple.NewTupleKey("document:a", "viewer", "user:anne"),
+		tuple.NewTupleKey("document:b", "viewer", "user:anne"),
+	}
+
+	for _, tupleKey := range tupleKeys {
+		resp, err := singleflightCheckResolver.ResolveCheck(ctx, &ResolveCheckRequest{
```

Comment:
Same comment as in https://github.com/openfga/openfga/pull/1376/files#r1493075019